_posts/2022-01-28-skyhook-bringing-computation-to-storage-with-apache-arrow.md

```diff
@@ -0,0 +1,219 @@
+Meanwhile, even though the storage system has its own compute capabilities, it’s relegated to just serving “dumb bytes”.
+
+Thanks to the [Center for Research in Open Source Software][cross] (CROSS) at the University of California, Santa Cruz, Apache Arrow 7.0.0 includes Skyhook, an [Arrow Datasets][dataset] extension that solves this problem by using the storage layer to reduce client resource utilization.
+We’ll examine the developments surrounding Skyhook as well as how Skyhook works.
+
+## Introducing Programmable Storage
+
+Skyhook is an example of programmable storage: exposing higher-level functionality from storage systems for clients to build upon.
+This allows us to make better use of existing resources (both hardware and development effort) in such systems, reduces the implementation burden of common operations for each client, and enables such operations to scale with the storage layer.
+
+Historically, big data systems like Apache Hadoop have tried to colocate computation and storage for efficiency.
+More recently, cloud and distributed computing have disaggregated computation and storage for flexibility and scalability, but at a performance cost.
+Programmable storage strikes a balance between these goals, allowing some operations to be run right next to the data while still keeping data and compute separate at a higher level.
+
+In particular, Skyhook builds on [Ceph][ceph], a distributed storage system that scales to exabytes of data while being reliable and flexible.
+With its Object Class SDK, Ceph enables programmable storage by allowing extensions that define new object types with custom functionality.
+
+## Skyhook Architecture
+
+Let’s look at how Skyhook applies these ideas.
+Overall, the idea is simple: the client should be able to ask Ceph to perform basic operations like decoding files, filtering the data, and selecting columns.
+That way, the work gets done using existing storage cluster resources, which means it’s both adjacent to the data and can scale with the cluster size.
+Also, this reduces the data transferred over the network, and of course reduces the client workload.
+
+On the storage system side, Skyhook uses the Ceph Object Class SDK to define scan operations on data stored in Parquet or Feather format.
+To implement these operations, Skyhook first implements a file system shim in Ceph’s object storage layer, then uses the existing filtering and projection capabilities of the Arrow Datasets library on top of that shim.
+
+Then, Skyhook defines a custom “file format” in the Arrow Datasets layer.
+Queries against such files get translated to direct requests to Ceph using those new operations, bypassing the traditional POSIX file system layer.
+After decoding, filtering, and projection, Ceph directly sends the client Arrow record batches, minimizing CPU overhead for encoding/decoding—another optimization Arrow makes possible.
```

Comment:
Rewrite? "After decoding, filtering, and projecting, Ceph sends the Arrow record batches directly to the client,..."